### ISSUE TITLE
Crashes when allocating memory

### ISSUE THREAD

**Issue by mathbr** [borgbackup/borg#521]
I'm seeing errors like this while trying to perform a backup:

```
hashindex: malloc buckets failed(Cannot allocate memory)
Local Exception.
Traceback (most recent call last):
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 1261, in main
    exit_code = archiver.run(args)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 1215, in run
    return args.func(args)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 183, in do_create
    read_special=args.read_special, dry_run=dry_run)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 247, in _process
    read_special=read_special, dry_run=dry_run)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 247, in _process
    read_special=read_special, dry_run=dry_run)
  File "/usr/lib/python3/dist-packages/borg/archiver.py", line 221, in _process
    status = archive.process_file(path, st, cache)
  File "/usr/lib/python3/dist-packages/borg/archive.py", line 535, in process_file
    chunks.append(cache.add_chunk(self.key.id_hash(chunk), chunk, self.stats))
  File "/usr/lib/python3/dist-packages/borg/cache.py", line 359, in add_chunk
    self.chunks[id] = (1, size, csize)
  File "borg/hashindex.pyx", line 174, in borg.hashindex.ChunkIndex.__setitem__ (borg/hashindex.c:3382)
Exception: hashindex_set failed
```

The directory to back up has 6441 files with sizes ranging from ~250MB to ~2GB.

I first tried with compression but gave up on that reading that it eats even more memory. My command line includes `--chunker-params 19,23,21,4095` but to no avail.

Anything else I could try?

**Comment by jungle-boogie** [comment]
Duplicate of this bug:
https://github.com/borgbackup/borg/issues/509

**Comment by mathbr** [comment]
@jungle-boogie Indeed looks like. I have 8GB of RAM and 2GB swap here BTW.

**Comment by ThomasWaldmann** [comment]
Likely it is duplicate, but still it would be good if you gave more complete information:
- commandline you used
- all output
- traceback
- system information, including borg version
- count and size of the files backed up in your repository (you can estimate memory usage yourself using the formula given in "internals" docs)

What you can try:
- add more memory, either swap (slow) or real
- work without files index (slow)

In case you used a smaller chunksize at the beginning: start a new repo with big chunksize from scratch.

**Comment by mathbr** [comment]
Sorry for not providing as much info as normally. I'm also using version 0.29 and already mentioned the relevant parts (traceback, `--chunker-params`, file count). What system info exactly do you want?

The actual file count is 6044 (according to `find -type f | wc -l`) and the total size is 3141604436 bytes (according to `df`). If I did the math correctly (and I'm not claiming I do) this gives me a memory usage of ~239 GB which I won't be able to achieve, no matter what. ;-)

I'll give it another try with a bigger initial chunk size (I indeed switched only after the first failure) and will add more info if it fails or works.

@ThomasWaldmann: Wouldn't adding that calculated memory usage estimate to `borg create --dry-run --stats ...` be  useful? That way people can check the expected memory usage and tweak the --chunker-params` before actually starting (possibly long running) backups.

**Comment by mathbr** [comment]
Note to myself: do not use a literal `^` in ipython. :-X Gotta recalculate the memory estimate with `**`

**Comment by dragetd** [comment]
df shows by default the amount of blocks, not bytes. 3141604436 bytes would be 3GB of data… a RAM usage of 8MB sounds correct. But 3141604436 4k-blocks means almost 12TB of data! Quiet a difference :-P
Initially you said 'files between 250MB and 2GB', so 10+ TB sounds realistic.

With so much data, you chunk index will be *huge*. But since you have relatively few files, increasing the chunk parameters to 21,23,22 or even 23,23,23 will help a lot. You can do the math to get better estimates.

**Comment by ThomasWaldmann** [comment]
@mathbr https://github.com/borgbackup/borg/issues/521#issuecomment-169094152 should be edited after you got the numbers right, so all followup comments can be deleted.

If you initially used a smaller chunksize before switching to a bigger one, you already have created a lot of chunks in your repository which need a lot of memory to deal with. You can have a look at .cache/borg/chunks file size or borg info output to get an impression.

**Comment by ThomasWaldmann** [comment]
see also #527.

**Comment by mathbr** [comment]
With the following command line everything worked out just fine now:

    borg create --progress --stats --chunker-params 19,23,21,4095 /.../Backup::2016-01-05 /.../Source

This time not even swap was used.

Here are the stats outputted afterwards:

```
------------------------------------------------------------------------------                                                                                                                              
Archive name: 2016-01-05
Archive fingerprint: e3e5b8b97983dfe9378712cf6597c693f76f8c1b3a9d381ff7bc55e9efb0990d
Start time: Tue Jan  5 19:57:56 2016
End time: Wed Jan  6 10:10:56 2016
Duration: 14 hours 13 minutes 0.57 seconds
Number of files: 6044

                       Original size      Compressed size    Deduplicated size
This archive:                3.22 TB              3.22 TB              3.21 TB
All archives:                3.22 TB              3.22 TB              3.21 TB

                       Unique chunks         Total chunks
Chunk index:                 1254688              1256701
------------------------------------------------------------------------------
```

It's unfortunate that basically no deduplication was performed. Any idea what I could tweak here and still keeps the memory usage sane?

**Comment by ThomasWaldmann** [comment]
Well, no dedup for this first backup likely means you do not have much "inner" duplication in the files you back up.

When you run the next backup(s) and your files didn't change much, you will see a huge "historical" deduplication.

If your data is compressible, you could add lz4 compression (which is very fast) or gzip compression (which is slower, but compresses a little better).

**Comment by mathbr** [comment]
@ThomasWaldmann Thanks, I'll give that a try.
